ratelimit.go

```diff
@@ -0,0 +1,157 @@
+	filters []Filter
+	// exceptions stores a list of exceptions to determine if should not apply the rate limiter.
+	exceptions []Exception
+}
+
+// NewTimeLimiter creates a new time based rate limiter middleware.
+func NewTimeLimiter(timeWindow time.Duration, capacity int64) *Limiter {
+	return &Limiter{
+		responder:     RateLimitResponder,
+		timeWindow:    timeWindow,
+		lastAvailable: time.Now(),
+		bucket:        ratelimit.NewBucket(timeWindow, capacity),
+	}
+}
+
+// NewRateLimiter creates a rate limiter middleware which limits the
+// amount of requests accepted per second.
+func NewRateLimiter(rate float64, capacity int64) *Limiter {
+	return &Limiter{
+		responder:  RateLimitResponder,
+		timeWindow: time.Second,
+		bucket:     ratelimit.NewBucketWithRate(rate, capacity),
+	}
+}
+
+// SetResponder sets a custom function to reply in case of rate limit reached.
+func (l *Limiter) SetResponder(fn http.HandlerFunc) {
+	l.responder = fn
+}
+
+// AddFilter adds a new rate limiter whitelist filter.
+// If the filter matches, the traffic won't be limited.
+func (l *Limiter) AddFilter(fn ...Filter) {
+	l.filters = append(l.filters, fn...)
+}
+
+// AddException adds a new rate limiter whitelist filter.
+// If the filter matches, the traffic won't be limited.
+func (l *Limiter) AddException(fn ...Exception) {
+	l.exceptions = append(l.exceptions, fn...)
+}
+
+// Register registers the middleware handler.
+func (l *Limiter) Register(mw layer.Middleware) {
+	mw.UsePriority("request", layer.TopHead, l.LimitHTTP)
+}
+
+// LimitHTTP limits an incoming HTTP request.
+// If some filter passes, the request won't be limited.
+// This method is used internally, but made public for public testing.
+func (l *Limiter) LimitHTTP(h http.Handler) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Run exceptions to ignore the limiter, if necessary
+		for _, exception := range l.exceptions {
+			if exception(r) {
+				h.ServeHTTP(w, r)
+				return
+			}
+		}
+
+		// Pass filters to determine if should apply the limiter.
+		// All the filtes must pass to apply the limiter.
+		for _, filter := range l.filters {
+			if !filter(r) {
+				h.ServeHTTP(w, r)
+				return
+			}
+		}
+
+		// Apply the rate limiter
+		available := l.bucket.TakeAvailable(1)
+
+		headers := w.Header()
+		headers.Set("X-RateLimit-Limit", strconv.Itoa(l.capacity()))
+		headers.Set("X-RateLimit-Remaining", strconv.Itoa(l.remaining()))
+
+		// If tokens are not available, reply with error, usually with 429
+		if available == 0 {
+			headers.Set("X-RateLimit-Reset", strconv.Itoa(int(l.resetTime())))
+			l.responder(w, r)
+			return
+		}
+
+		// Otherwise track time and forward the request
+		l.trackTime()
+		h.ServeHTTP(w, r)
+	}
+}
+
+// resetTime is used to calculate the pending reset time to wait.
+func (l *Limiter) resetTime() time.Duration {
+	l.lm.Lock()
+	defer l.lm.Unlock()
+	return time.Now().Sub(l.lastAvailable) / time.Second
```

Comment:
Well, the idea is basically to calculate the pending wait time until the bucket has available tokens. 
My initial approach is not good, actually. I'll rethink about this.